### PR TITLE
[VEN-1146]: Added vip for comptroller diamond proxy

### PR DESCRIPTION
**Pull request by Debugger022**
## Description

Resolves
- VIP for the Comptroller diamond proxy(testnet).

### CODE REVIEW

File: simulations/vip-diamond-comptroller-testnet/simulations.ts [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+        venusBorrowState = await unitroller.venusBorrowState(BUSD.address);
+
+        venusAccrued = await unitroller.venusAccrued(BUSD.address);
+
+        vaiMintRate = await unitroller.vaiMintRate();
+
+        vaiController = await unitroller.vaiController();
+
+        mintedVAIs = await unitroller.mintedVAIs(busdHolder.address);
+        unitroller.minte;
```
Comment (chechu) [inline simulations/vip-diamond-comptroller-testnet/simulations.ts:148]: ?

File: simulations/vip-diamond-comptroller-testnet/simulations.ts [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+
+        // cheking all public mappings
+        market = await unitroller.markets(vBUSD.address);
+
+        venusBorrowerIndex = await unitroller.venusBorrowerIndex(vBUSD.address, busdHolder.address);
+
+        venusSupplierIndex = await unitroller.venusSupplierIndex(vBUSD.address, busdHolder.address);
+
+        venusBorrowSpeeds = await unitroller.venusBorrowSpeeds(vUSDT.address);
+        venusSupplySpeeds = await unitroller.venusSupplySpeeds(vUSDT.address);
```
Comment (chechu) [inline simulations/vip-diamond-comptroller-testnet/simulations.ts:182]: I would do this for every market returned by `unitroller.getAllMarkets`, not only for vBUSD and vUSDT

File: simulations/vip-diamond-comptroller-testnet/simulations.ts [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+      });
+    });
+  });
+
+  testVip("VIP-Diamond TRON Contract Migration", vipDiamondTestnet());
```
Comment (chechu) [inline simulations/vip-diamond-comptroller-testnet/simulations.ts:187]:
```suggestion
  testVip("VIP-Diamond Contract Migration", vipDiamondTestnet());
```

File: simulations/vip-diamond-comptroller-testnet/simulations.ts [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+
+      it("setting close factor", async () => {
+        const currentCloseFactor = (await diamondUnitroller.closeFactorMantissa()).toString();
+        await diamondUnitroller.connect(owner)._setCloseFactor(parseUnits("10000", 18));
+        expect(await diamondUnitroller.closeFactorMantissa()).to.equals(parseUnits("10000", 18));
+        await diamondUnitroller.connect(owner)._setCloseFactor(parseUnits(currentCloseFactor, 0));
+        expect(await diamondUnitroller.closeFactorMantissa()).to.equals(parseUnits(currentCloseFactor, 0));
+      });
+
+      it("setting setting Liquidation Incentive", async () => {
```
Comment (chechu) [inline simulations/vip-diamond-comptroller-testnet/simulations.ts:323]:
```suggestion
      it("setting Liquidation Incentive", async () => {
```

File: simulations/vip-diamond-comptroller-testnet/simulations.ts (hunk 1 above)
Comment (defcon022) [reply simulations/vip-diamond-comptroller-testnet/simulations.ts]: Removed, cfe4633bf5ca7d1e1d013d6b9dcd7206e3050a02

File: simulations/vip-diamond-comptroller-testnet/simulations.ts (hunk 2 above)
Comment (defcon022) [reply simulations/vip-diamond-comptroller-testnet/simulations.ts]: Added for all markets, cfe4633bf5ca7d1e1d013d6b9dcd7206e3050a02

File: simulations/vip-diamond-comptroller-testnet/simulations.ts (hunk 3 above)
Comment (defcon022) [reply simulations/vip-diamond-comptroller-testnet/simulations.ts]: Fixed, cfe4633bf5ca7d1e1d013d6b9dcd7206e3050a02

**Comment by Debugger022** [comment]
Changes in [this commit](https://github.com/VenusProtocol/vips/pull/11/commits/6cdc98b1677e6e0c8abfdc2726515b4e2063abfd) are pushed due to the changes in [this commit](https://github.com/VenusProtocol/venus-protocol/pull/312/commits/cfaa69aea6ef1d55c7fc4e457780ca83cd58add1)

File: vips/vip-Diamond-comptroller-testnet.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      {
+        target: ACM,
+        signature: "revokeCallPermission(address,string,address)",
+        params: [UNITROLLER, "_setActionsPaused(address[],uint256[],bool)", NORMAL_TIMELOCK],
+      },
+      {
+        target: ACM,
+        signature: "giveCallPermission(address,string,address)",
+        params: [UNITROLLER, "_setActionsPaused(address[],uint8[],bool)", NORMAL_TIMELOCK],
+      },
```
Comment (kkirka) [inline vips/vip-Diamond-comptroller-testnet.ts:48]: What about other timelocks and the guardian multisig? I think we should we update their permissions as well here, at least it's what I see in the access list sheet (see "Governance Improvements" in Notion), and we should update the access list sheet with the new signature.

File: simulations/vip-diamond-comptroller-testnet/simulations.ts [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+import { vipDiamondTestnet } from "../../vips/vip-Diamond-comptroller-testnet";
+import Comptroller from "./abi/Comptroller.json";
+import IERC20Upgradeable from "./abi/IERC20UpgradableAbi.json";
+import VBEP20_DELEGATE_ABI from "./abi/VBep20DelegateAbi.json";
+
+const UNITROLLER = "0x94d1820b2D1c7c7452A163983Dc888CEC546b77D";
+const DIAMOND = "0xB2243Da976F2cbAAa4dd1a76BF7F6EFbe22c4CFc";
+
+const Owner = "0xce10739590001705F7FF231611ba4A48B2820327";
+const zeroAddr = "0x0000000000000000000000000000000000000000";
```
Comment (kkirka) [inline simulations/vip-diamond-comptroller-testnet/simulations.ts:17]:
```suggestion
const zeroAddr = ethers.constants.AddressZero;
```

File: simulations/vip-diamond-comptroller-testnet/simulations.ts [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+const VBUSD = "0x08e0A5575De71037aE36AbfAfb516595fE68e5e4";
+const VUSDT = "0xb7526572FFE56AB9D7489838Bf2E18e3323b441A";
+
+const initMainnetUser = async (user: string) => {
+  await impersonateAccount(user);
+  return ethers.getSigner(user);
+};
```
Comment (kkirka) [inline simulations/vip-diamond-comptroller-testnet/simulations.ts:24]: Maybe we could reuse the existing one?

```suggestion
import { initMainnetUser } from "../../src/utils";
```

**Review by kkirka — COMMENTED** [review]
LGTM, the upgrade is quite thoroughly tested. With storage layout and mint/borrow/repay/redeem tests I'm quite confident the upgrade will go smoothly <3

File: vips/vip-Diamond-comptroller-testnet.ts (hunk 5 above)
Comment (Debugger022) [reply vips/vip-Diamond-comptroller-testnet.ts]: Added, be0cf490a06d3861443b1bfd6bb6e4ffdd6c1b5f

Comment (Debugger022) [reply vips/vip-Diamond-comptroller-testnet.ts]: @kkirka are you ok with the above fix?

**Review by GitGuru7 — APPROVED** [review]

File: simulations/vip-diamond-comptroller/vip-diamond-comptroller/utils/cut-params.json [hunk 8]
```diff
@@ -0,0 +1,3 @@
+{
+  "cutParams": []
```
Comment (chechu) [inline simulations/vip-diamond-comptroller/vip-diamond-comptroller/utils/cut-params.json:2]: We need this

Comment (Debugger022) [reply simulations/vip-diamond-comptroller/vip-diamond-comptroller/utils/cut-params.json:2]: yes will be adding it. Once the Diamond for mainnet will be deployed.

**Review by chechu — APPROVED** [review]

File: vips/vip-Diamond-comptroller-testnet.ts (hunk 5 above)
Comment (kkirka) [reply vips/vip-Diamond-comptroller-testnet.ts]: LGTM

**Comment by chechu** [comment]
https://app.venus.io/#/governance/proposal/174

**Pull request merged**